crates/core/c8y_api/src/json_c8y.rs

```diff
@@ -236,95 +238,167 @@ impl SourceInfo {
-            alarm_type,
-            time,
-            text,
-            fragments,
+impl From<&ThinEdgeAlarm> for C8yCreateAlarm {
+    fn from(alarm: &ThinEdgeAlarm) -> Self {
+        let source = Self::convert_source(alarm);
+        let alarm_type = Self::convert_alarm_type(alarm);
+
+        match alarm.data.as_ref() {
+            None => Self {
+                alarm_type,
+                source,
+                data: None,
+            },
+            Some(tedge_alarm_data) => {
+                let alarm_data = C8yCreateAlarmData {
+                    severity: Self::convert_severity(tedge_alarm_data),
+                    text: Self::convert_text(tedge_alarm_data, &alarm_type),
+                    time: Self::convert_time(tedge_alarm_data),
+                    fragments: Self::convert_extras(tedge_alarm_data),
+                };
+                Self {
+                    source,
+                    alarm_type,
+                    data: Some(alarm_data),
+                }
+            }
         }
     }
 }
 
-impl TryFrom<&ThinEdgeAlarm> for C8yCreateAlarm {
-    type Error = SMCumulocityMapperError;
-
-    fn try_from(alarm: &ThinEdgeAlarm) -> Result<Self, SMCumulocityMapperError> {
-        let severity = alarm.severity.to_string();
-        let alarm_type = alarm.name.to_owned();
-        let text;
-        let time;
-        let fragments;
-
-        match &alarm.to_owned().data {
-            None => {
-                text = alarm_type.clone();
-                time = OffsetDateTime::now_utc();
-                fragments = HashMap::new();
-            }
-            Some(data) => {
-                text = data.text.clone().unwrap_or_else(|| alarm_type.clone());
-                time = data.time.unwrap_or_else(OffsetDateTime::now_utc);
-                fragments = data.alarm_data.clone();
+impl C8yCreateAlarm {
+    fn convert_source(alarm: &ThinEdgeAlarm) -> Option<SourceInfo> {
+        match alarm.source.r#type {
+            EntityType::MainDevice => None,
+            EntityType::ChildDevice => {
+                Some(make_c8y_source_fragment(&alarm.source.entity_id.clone()))
             }
+            EntityType::Service => None, // TODO!: What is for service?
         }
+    }
 
-        let source = if let Some(external_source) = &alarm.source {
-            make_c8y_source_fragment(external_source)
-        } else {
-            None
-        };
+    fn convert_alarm_type(alarm: &ThinEdgeAlarm) -> String {
+        match alarm.alarm_type.clone() {
+            Some(alarm_type) if !alarm_type.is_empty() => alarm_type,
+            _ => DEFAULT_ALARM_TYPE.to_string(),
+        }
+    }
 
-        Ok(Self {
-            source,
-            severity,
-            alarm_type,
-            time,
-            text,
-            fragments,
-        })
+    // TODO: Ask which one is better, using default or returning an error.
+    fn convert_severity(alarm_data: &ThinEdgeAlarmData) -> AlarmSeverity {
```

Comment:
In that case, using a default value is appropriate and convenient for the users.